Implement mute user feature.

abhijeetbodas2001 opened the pull request:
#8040

Implementation plan at https://chat.zulip.org/api/mute-user

### Remaining tasks
- [ ] Hide 1:1 PMs from muted user everywhere
- [x] Stream messages and huddle - tuck away under "Click to reveal" dialogue
- [x] Exclude muted users from typeaheads
- [x] Exclude muted users from typing status
- [ ] Make UI for muting/unmuting users available in production, and add a /help document.


<details><summary>PRs and commits which were merged as parts of this feature</summary>

- Commits for API work → 89f61395059788fe7c43bfd023b285441ee22370, 3bfcaa39689c9996cb7eb0ec65bd37f727074f6e, 1de9444242950d0f54e1e554aeb6e400ef8ad474
- Follow up work from 3bfcaa39689c9996cb7eb0ec65bd37f727074f6e → #18045
- All backend commits → #18086
- Buddy list work → #18218
- UI work (dev-only) → #18233 

</details>

abhijeetbodas2001 commented:
@zulipbot remove "has conflicts"

abhijeetbodas2001 commented:
@showell @amanagr This PR is ready for review.

abhijeetbodas2001 commented:
@showell @amanagr This is ready for review.

abhijeetbodas2001 commented:
@showell @timabbott This is ready for review.

abhijeetbodas2001 commented:
@ryanreh99 As someone who's worked on topic muting as well as message lists before, could you review this?

ryanreh99 reviewed (commented):
Hi, I haven't gone through all the commits yet but I played with it and it looks really nice.
I left a few comments.

abhijeetbodas2001 commented:
Thanks for reviewing! I'll take a close look at the comments you posted and get back in a few days. I'm currently a bit busy with college exams.

abhijeetbodas2001 commented on static/js/click_handlers.js:
```diff
@@ -211,6 +211,16 @@ exports.initialize = function () {
+        const message_id = rows.id_for_recipient_row(row);
+        const message = current_msg_list.get(message_id);
+        current_msg_list.view.rerender_messages([message], false, true);
+        if (current_msg_list === message_list.narrowed) {
+            home_msg_list.view.rerender_messages([message], false, true);
+        }
```
Thanks for pointing this out!
I haven't discussed this point specifically with anyone or on CZO, but this probably isn't intended.
The behaviour currently is-
1. In a narrow, reveal the message.
2. Click on the narrow again from the left sidebar, which will rerender it, and the message will be "hidden" again.
3. Click on "All messages", but here, the message is still "revealed"

and, independently,
1. In the home view, reveal the message
2. Narrow to some topic/stream - it is "hidden" again
3. Go back to the home view - it is still "reveal"-ed there.

Since we don't rerender the home view, as you pointed out, the temporary "reveal" mechanism breaks ☹️ 
I think this is what could be done here-
1. On "reveal"-ing from a narrow - reveal the message in the narrow only, and not in the home view. This will work as expected.
2. On "reveal"-ing from the home view, - we can either
  i.  Not do anything on click. This will feel weird to the user, but will prevent the message from staying "reveal"-ed in the home view until the next tab refresh.
  ii. Reveal the message. It will stay revealed in the home view, but users won't feel like the button is broken.

I think we should definitely do (1), but I'm not sure about which option to go with between 2(i) and 2(ii). What are your thoughts on this?

abhijeetbodas2001 replied:
https://chat.zulip.org/#narrow/stream/2-general/topic/Mute.20user.20.2316915/near/1123931

abhijeetbodas2001 commented on static/js/muting.js:
```diff
@@ -72,6 +73,46 @@ exports.set_muted_topics = function (tuples) {
+};
+
+exports.remove_muted_user = function (user_id) {
+    if (user_id) {
```
There aren't any calls to `remove_muted_user` which I know can pass `undefined`. And it looks like javascript doesn't complain on trying to delete `undefined` key either. Do you suggest removing this check?

abhijeetbodas2001 commented:
@ryanreh99 Thanks for the review. In the last push, I rebased on top of master, addressed two of your comments and replied on the other two.

timabbott commented on zerver/views/muting.py:
```diff
@@ -85,3 +86,29 @@ def update_muted_topic(
+    if user_profile.id == muted_user_id:
+        return json_error(_("Cannot mute self"))
+
+    muted_user_profile = get_user_profile_by_id(muted_user_id)
```
This should be `access_user_by_id`, same for the other endpoint -- otherwise you can access a random user ID not in your realm.

abhijeetbodas2001 replied:
Done. Changed to `access_user_by_id`.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -1819,14 +1819,19 @@ def do_send_messages(
+
+            # Extend the list with users who have muted the sender.
+            mark_as_read_msg = list(mark_as_read)
+            mark_as_read_msg.extend(get_muting_users(send_request.message.sender))
```
I think it'd be worth using `Set` here; e.g. `get_muting_users` could return a set, and then you could just add everything in the `mark_as_read` parameter to it (avoiding mutating that parameter).

abhijeetbodas2001 replied:
This now uses a `Set`, just like other inputs to `create_user_messages`.

timabbott commented on static/js/message_events.js:
```diff
@@ -393,11 +393,11 @@ export function update_messages(events) {
         // we just changed the narrow earlier in this function.
         if (!changed_narrow && current_msg_list === message_list.narrowed) {
-            message_list.narrowed.update_muting_and_rerender();
+            message_list.narrowed.update_and_rerender_for_topic_muting();
```
Can you pull this commit out in a separate PR (or just move it to be the first commit on the branch)?  We can merge it right now, since it's a helpful change in any case.

(Though I think I'd prefer the name `update_topic_muting_and_rerender`).

abhijeetbodas2001 replied:
Shifted this commit to the beginning of the PR, and also pulled out a couple other refactoring changes from the main commits 🙂

timabbott commented:
For the settings UI, I think a separate "Muted users" top-level page (keeping the topics one as "Muted topics") is probably cleaner; one never wants to have two tables on top of each other where if the top table were really big, one might not notice they both exist.

abhijeetbodas2001 commented:
@timabbott Thanks for the review! I've updated the PR, and it's ready for another look. Here are the changes-
- Shifted all refactoring commits to the beginning of the PR, and also slid in a couple of bug fixes with the recent topics view which I noticed.
- Made a separate settings page for showing muted users list. I wasn't sure of what icon to use, though. I've used `fa-eye-slash` currently. Screenshots in the main PR description above.
- `get_muting_users` now returns a `Set` of muter IDs, which is then extended and passed to `create_user_messages`.
- Switched the backend views code to use `access_user_by_id` instead of `get_user_profile_by_id`. I wasn't sure whether we should allow muting **bots**, because the use case for muting bots could be different than users. Would be great if you could shed some light on this. The current implementation *does not allow* muting bots, in `access_user_by_id` as well as `user-info-popovers`s.

timabbott commented:
I merged the first couple commits as the series ending with f725711ff279c2aaca7c79d02a5a2fb6688966e3, after fixing a few typos in the first one's commit message (mainly s/recieve/receive/, which is one of those sneaky ones).

timabbott commented:
OK, and I merged the remaining preparatory refactors without changes as the series ending with 95663693dfa2034598507f8b6dd9bf5b5afd3284.  Thanks @abhijeetbodas2001!

I think fa-eye-slash is a reasonable icon for muting users; that's an easy thing to iterate on later holistically.  

I think the next step for this is probably for @alya and I to play with it a bit.

timabbott commented:
This is great progress!  I've merged the first 3 commits as the series ending with 1de9444242950d0f54e1e554aeb6e400ef8ad474, after making several changes to the main backend commit.
* Renamed the `muted_user_profile` field to `muted_user` for consistency with other foreign keys.
* Attempted the backend change for `import_realm`.  I added what I think is the right code:
```
+    if "zerver_muteduser" in data:
+        fix_datetime_fields(data, "zerver_muteduser")
+        re_map_foreign_keys(data, "zerver_muteduser", "user_profile", related_table="user_profile")
+        re_map_foreign_keys(data, "zerver_muteduser", "muted_user", related_table="user_profile")
+        update_model_ids(MutedUser, data, "muteduser")
+        bulk_import_model(data, MutedUser)
+
```
but we should really have test coverage in `test_import_export.py`, so please add that in your next push to this PR.  (Look at MutedTopic testing for an example; and @hackerkid can help advise).

* I applied these additional changes to the backend commit.  Most are self-explanatory; for the API documentation, I cut content because I think we'll want to link to a /help/ article explaining what muting users does once such links won't be broken; can you add links to the commit that adds /help/ articles?
```
diff --git a/templates/zerver/api/unmute-user.md b/templates/zerver/api/unmute-user.md
index 9ff042c212..7e0706e6c7 100644
--- a/templates/zerver/api/unmute-user.md
+++ b/templates/zerver/api/unmute-user.md
@@ -27,6 +27,6 @@ A typical successful JSON response may look like:
 
 {generate_code_example|/users/me/muted_users/{muted_user_id}:delete|fixture(200)}
 
-An example JSON response for when a user is not previously muted
+An example JSON response for when a user is not previously muted:
 
 {generate_code_example|/users/me/muted_users/{muted_user_id}:delete|fixture(400)}
diff --git a/zerver/lib/user_mutes.py b/zerver/lib/user_mutes.py
index c1e053d500..6997e2beaf 100644
--- a/zerver/lib/user_mutes.py
+++ b/zerver/lib/user_mutes.py
@@ -8,7 +8,7 @@ from zerver.models import MutedUser, UserProfile
 
 
 def get_user_mutes(user_profile: UserProfile) -> List[Dict[str, int]]:
-    rows = MutedUser.objects.filter(user_profile=user_profile,).values(
+    rows = MutedUser.objects.filter(user_profile=user_profile).values(
         "muted_user_profile__id",
         "date_muted",
     )
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 5083342e45..f77e856e0f 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -5639,12 +5639,8 @@ paths:
       operationId: mute_user
       tags: ["users"]
       description: |
-        This endpoint mutes a user.
-        As soon as a muted user sends a message which the user can receive,
-        it will be marked as read for the user, and hence will not contribute
-        to any notifications or unread counters.
-        Any old unread messages will also be marked as read when a user
-        is muted.
+        This endpoint mutes a user.  Messages sent by users you've muted will
+        be automatically marked as read and hidden.
 
         `POST {{ api_url }}/v1/users/me/muted_users/{muted_user_id}`
 
diff --git a/zerver/tests/test_muting_users.py b/zerver/tests/test_muting_users.py
index 09aeb88bb6..ef2c61261f 100644
--- a/zerver/tests/test_muting_users.py
+++ b/zerver/tests/test_muting_users.py
@@ -13,10 +13,11 @@ class MutedUsersTests(ZulipTestCase):
 
         muted_users = get_user_mutes(othello)
         self.assertEqual(muted_users, [])
+        mute_time = datetime(2021, 1, 1, tzinfo=timezone.utc)
 
         with mock.patch(
             "zerver.lib.user_mutes.timezone_now",
-            return_value=datetime(2021, 1, 1, tzinfo=timezone.utc),
+            return_value=mute_time,
         ):
             add_user_mute(user_profile=othello, muted_user_profile=cordelia)
 
@@ -27,7 +28,7 @@ class MutedUsersTests(ZulipTestCase):
             muted_users[0],
             {
                 "id": cordelia.id,
-                "timestamp": datetime_to_timestamp(datetime(2021, 1, 1, tzinfo=timezone.utc)),
+                "timestamp": datetime_to_timestamp(mute_time),
             },
         )
 
diff --git a/zerver/views/muting.py b/zerver/views/muting.py
index be28b5b1c0..dce0ace8bb 100644
--- a/zerver/views/muting.py
+++ b/zerver/views/muting.py
@@ -93,7 +93,9 @@ def mute_user(request: HttpRequest, user_profile: UserProfile, muted_user_id: in
     if user_profile.id == muted_user_id:
         return json_error(_("Cannot mute self"))
 
-    muted_user_profile = access_user_by_id(user_profile, muted_user_id, for_admin=False)
+    muted_user_profile = access_user_by_id(
+        user_profile, muted_user_id, allow_bots=False, for_admin=False
+    )
     date_muted = timezone_now()
 
     if user_is_muted(user_profile, muted_user_profile):
@@ -106,7 +108,9 @@ def mute_user(request: HttpRequest, user_profile: UserProfile, muted_user_id: in
 def unmute_user(
     request: HttpRequest, user_profile: UserProfile, muted_user_id: int
 ) -> HttpResponse:
-    muted_user_profile = access_user_by_id(user_profile, muted_user_id, for_admin=False)
+    muted_user_profile = access_user_by_id(
+        user_profile, muted_user_id, allow_bots=False, for_admin=False
+    )
 
     if not user_is_muted(user_profile, muted_user_profile):
         return json_error(_("User is not muted"))
```

I also pushed back to this PR, since I took care of rebasing for the `muted_user_profile => muted_user` rename while testing.

Here's some notes on additional follow-up work from the merged commits:
* Link to /help/ for muted users from API docs for the feature, both the main docs and the events.
* We should change the `do_` functions to record entries in RealmAuditLog when muting another user.  I think `modified_user` needs to be the current user, as well as `acting_user`, so probably we need to store the other user in `extra_data` (no need to record timestamp, since that's there anyway).
* For unmute_user, would it be cleaner to pass the `MutedUser` object from the `user_is_muted` check to the `remove_user_mute` function?  This would avoid a second fetch.
* The backend tests should be tweaked to (1) do everything via the API, rather than calling `do_` functions directly -- it's better to have tests be more end-to-end, and (2) do all the repeated times in variables.  (I changed a few bits in this direction before merging, but decided to leave the rest to you).

A few more high-level thoughts on the UI:
* The label for the popover "Mute" button should be clearer, maybe "Mute this user" is good enough for now.
* Medium-term, we'll want to redo the styling for hiding messages from a hidden user to also hide the user's avatar and be more nicely visually designed.  But I think that can wait until this integrated, since at first this will be a rarely used feature, so it's not a big deal if the visuals aren't perfect at first.

I'll post a series of comments on the next few commits now :)

timabbott commented on static/js/muting_ui.js:
```diff
@@ -158,6 +163,59 @@ export function toggle_topic_mute(message) {
+    popovers.hide_user_info_popover();
+    popovers.hide_message_info_popover();
+    popovers.hide_user_sidebar_popover();
+    if (timestamp_ms() < last_user_update + 1000) {
```
This 1000 hack copied from topic-muting seems like a bug there as well.  I don't want to particularly block this on fixing that, but if you have time, I don't think the Zulip API ever has a use case for 1s timeouts like this, and we should figure out what other bug the topic_mute code there is masking and fix that (and then remove it).

abhijeetbodas2001 replied:
I opened #18279 for this.

timabbott commented on static/js/muting_ui.js:
```diff
@@ -158,6 +163,59 @@ export function toggle_topic_mute(message) {
+
+export function mute_user(user_id) {
+    popovers.hide_user_info_popover();
+    muting.add_muted_user(user_id);
```
It looks like this line is effectively local-echoing the user mute; I think they're be rare enough that it'd probably be better to just not do that sort of local echo (i.e. delete this line and let us get the update via `handle_user_updates` -- I think that'd let us get rid of the 1000 hack at least for this new code).

(Topic edits are really common and have a reasonable case for doing so).

timabbott commented on static/templates/confirm_mute_user.hbs:
```diff
@@ -0,0 +1,5 @@
+<p>
+    {{#tr this}}
+    Would you like to mute user <b>__user_name__</b>?  This will mark all received messages from __user_name__ as read, and subsequent messages sent by __user_name__ while muted will be marked as read automatically.
```
We may want to avoid mentioning the user's name 3 times; e.g.:
```
Would you like to mute user <b>__user_name__</b>?  Messages sent by muted users will never trigger notifications, will be marked as read, and will be hidden."
```

(We should also use a `help_link_widget` somewhere on this dialog to link to the /help/ article).

abhijeetbodas2001 replied:
Makes sends, done.

timabbott commented on static/templates/user_info_popover_content.hbs:
```diff
@@ -150,4 +150,23 @@
+    {{#if can_mute }}
+    <hr>
+    <li>
+        <a href="#" class="sidebar-popover-mute-user">
```
I think we want this link to not have an `href`, rather than having `"#"` as the href, after looking at other similar links in this file.  (Fixing this will prevent the browser from showing a URL when hovering this link, which is correct as no URL corresponds to this action).

abhijeetbodas2001 replied:
Fixed.

timabbott commented on templates/zerver/help/mute-a-user.md:
```diff
@@ -0,0 +1,47 @@
+don't contribute to unread counts on the left sidebar,
+and are hidden from the message view.
+
+Muted users don't appear in the right sidebar, and
```
"don't appear in the right sidebar or the [private messages](...) list".

abhijeetbodas2001 replied:
Ok, done.

timabbott commented on static/js/muting_ui.js:
```diff
@@ -158,6 +163,59 @@ export function toggle_topic_mute(message) {
+        url: "/json/users/me/muted_users/" + user_id,
+        idempotent: true,
+    });
+}
```
Structurally, it would be better for this commit to be after the commits that implement the actual hiding behavior in this PR.  While this commit is mostly complete, we won't want to merge a commit claiming that muting users does something, when it doesn't yet.

timabbott commented on static/js/settings_muted_users.js:
```diff
@@ -0,0 +1,24 @@
+        const $row = $(this).closest("tr");
+        const user_id = Number.parseInt($row.attr("data-user-id"), 10);
+
+        e.stopImmediatePropagation();
```
Why this rather than `.stopPropagation`?  I guess we do that in "Muted topics", but that's probably an accident.

abhijeetbodas2001 replied:
#18279 takes care of this in `settings/muted-topics`

timabbott commented on static/js/muting_ui.js:
```diff
@@ -187,11 +195,41 @@ export function handle_user_updates(muted_user_ids) {
+        parent_container: $("#muted-user-settings"),
+        simplebar_container: $("#muted-user-settings .progressive-table-wrapper"),
+    });
 }
```
This function more naturally belongs in `settings_muted_users.js`; can you move it?

And maybe add a prep commit to move the similar function to `settings_muted_topics.js`.

timabbott replied:
I think `rerender_on_user_update` has the same story.

abhijeetbodas2001 replied:
> And maybe add a prep commit to move the similar function to settings_muted_topics.js.

#18279 handles this.

> I think rerender_on_user_update has the same story.

Not really, that function handles a lot more stuff, not just the settings page.

timabbott commented on static/js/settings_muted_users.js:
```diff
@@ -0,0 +1,24 @@
+        e.stopImmediatePropagation();
+
+        muting_ui.unmute_user(user_id);
+        $row.remove();
```
I don't think we should do this UI update here of removing the row -- let's just let the live-update handle it.

Structurally, we have to make live-update work because you can have multiple browser tabs open and looking at this screen (Worth testing that BTW), and so generally it's better for buttons like this to only send the request to the server (and maybe do a loading spinner, if useful), and then have the success handler for that HTTP request just hide the loading spinner.  

Anything in the existing codebase that doesn't have that structure and isn't local echo for message feed is legacy code that we've neglected to update.

timabbott commented on frontend_tests/node_tests/popovers.js:
```diff
@@ -156,6 +156,7 @@ test_ui("sender_hover", (override) => {
                     can_set_away: false,
                     can_revoke_away: false,
                     can_mute: true,
+                    can_unmute: false,
```
This change probably belongs in the previous commit?

timabbott commented on static/templates/settings/muted_users_settings.hbs:
```diff
@@ -0,0 +1,13 @@
+<div id="muted-user-settings" class="settings-section" data-name="muted-users">
+    <input id="muted_users_search" class="search" type="text" placeholder="{{t 'Search muted users...' }}" aria-label="{{t 'Search muted users...' }}"/>
```
We apparently have several of this bug, but we should use …, aka the unicode `...` character, not 3 dots, since it's visibly prettier.  This is not new with your PR -- but maybe another opportunity to do a quick prep commit (or just separate mini PR) to clean up.
```
p$ git grep '[.][.][.]' static/templates/
static/templates/admin_user_group_list.hbs:        <div class="input" contenteditable="true" data-placeholder="{{t 'Add member...' }}"></div>
static/templates/message_body.hbs:<div class="message_expander message_length_controller" title="{{t 'Expand message (-)' }}">{{t "[More...]" }}</div>
static/templates/settings/attachments_settings.hbs:    <input id="upload_file_search" class="search" type="text" placeholder="{{t 'Search uploads...' }}" aria-label="{{t 'Search uploads...' }}"/>
static/templates/settings/muted_topics_settings.hbs:    <input id="muted_topics_search" class="search" type="text" placeholder="{{t 'Search muted topics...' }}" aria-label="{{t 'Search muted topics...' }}"/>
static/templates/settings/muted_users_settings.hbs:    <input id="muted_users_search" class="search" type="text" placeholder="{{t 'Search muted users...' }}" aria-label="{{t 'Search muted users...' }}"/>
static/templates/typing_notification.hbs:<li data-email="{{this.email}}" class="typing_notification">{{#tr this}}__full_name__ is typing...{{/tr}}</li>
```

abhijeetbodas2001 replied:
I'm going to change this text to "Filter muted users" in the spirit of b40e50f2959998d0193b65a3e5b4065ff5abfea5, but I've opened #18201 for the remaining ones.

timabbott replied:
Sounds great!

timabbott commented on zerver/lib/cache.py:
```diff
@@ -534,6 +534,10 @@ def realm_user_dicts_cache_key(realm_id: int) -> str:
     return f"realm_user_dicts:{realm_id}"
 
 
+def get_muting_users_cache_key(muted_user_profile: "UserProfile") -> str:
```
I guess I didn't update this commit for the `muted_user_profile => muted_user` rename; can you take care of that?

timabbott commented on zerver/lib/user_mutes.py:
```diff
@@ -41,3 +42,15 @@ def remove_user_mute(user_profile: UserProfile, muted_user: UserProfile) -> None
+    rows = MutedUser.objects.filter(
+        muted_user=muted_user,
+    ).values("user_profile__id")
+    return {row["user_profile__id"] for row in rows}
```
Can you get the raw SQL for this query and do an `EXPLAIN ANALYZE` on it in a `manage.py dbshell` (postgres shell) to make sure the query plan is an index scan?  You should be able to grab the generated SQL via the `queries_captured` context manager and just printing the queries rather than checking the length.

timabbott commented on zerver/lib/user_mutes.py:
```diff
@@ -41,3 +42,15 @@ def remove_user_mute(user_profile: UserProfile, muted_user: UserProfile) -> None
+    # This is kind of the inverse of `get_user_mutes` above.
+    # While `get_user_mutes` is mainly used for event system work,
+    # this is used in the message send codepath, to get a list
+    # of IDs of users who have muted a particular user.
```
You can make this comment a docstring, and that'd probably be nicer.  Also, add another sentence mentioning that this will include deactivated users.  (I think that behavior is probably correct, given how we'll use this).

You can also move this commit to be a prep commit on the branch, since it should be mergeable independently.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2018,7 +2023,7 @@ def create_user_messages(
     stream_email_user_ids: AbstractSet[int],
     mentioned_user_ids: AbstractSet[int],
-    mark_as_read: Sequence[int] = [],
+    mark_as_read_for_users: AbstractSet[int],
```
Why AbstractSet, and not Set?

timabbott commented on zerver/tests/test_muting_users.py:
```diff
@@ -153,3 +154,31 @@ def test_get_muting_users(self) -> None:
         self.assertEqual(set(), get_muting_users(muted_user))
         self.assertEqual(set(), cache_get(get_muting_users_cache_key(muted_user))[0])
+
+    def test_messages_from_muted_user_marked_as_read(self) -> None:
```
Let's call this `test_new_messages...` -- aka add `new`

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5036,6 +5036,52 @@ def do_mark_stream_messages_as_read(
     return count
 
 
+def do_mark_user_messages_as_read(
```
`do_mark_muted_user_messages_as_read` would be a better name.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5036,6 +5036,52 @@ def do_mark_stream_messages_as_read(
+
+    msgs = msgs.extra(
+        where=[UserMessage.where_unread()],
+    )
```
I'd prefer that these first several lines were a single block rather than reassigning the variable several times; also, we should call the variable `messages` -- our naming convention is to avoid abbreviations like `msgs`.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5036,6 +5036,52 @@ def do_mark_stream_messages_as_read(
+        messages=message_ids,
+        all=False,
+    )
+    event_time = timezone_now()
```
It would be a nice preparatory cleanup to create a `dataclass` for the `update_message_flags` event type so we stop duplicating these dictionaries, but would also be a fine followup, so feel free to just copy this idea onto your TODO list and focus on the rest.

timabbott commented on zerver/tests/test_muting_users.py:
```diff
@@ -182,3 +182,16 @@ def test_messages_from_muted_user_marked_as_read(self) -> None:
+
+        pm_to_iago = self.send_personal_message(muted_user, iago, "Spam in PM")
+
+        url = "/api/v1/users/me/muted_users/{}".format(muted_user.id)
```
Add a comment before this line explaining what API call we're about to do / what we're about to test.

timabbott commented on zerver/views/muting.py:
```diff
@@ -100,6 +106,8 @@ def mute_user(request: HttpRequest, user_profile: UserProfile, muted_user_id: in
         return json_error(_("User already muted"))
 
     do_mute_user(user_profile, muted_user, date_muted)
+    do_mark_user_messages_as_read(user_profile, muted_user)
```
Let's make `do_mute_user` call this internally?  That's safer, since it handles the case where some other code path starts being able to mute users as well.

timabbott commented on static/js/activity.js:
```diff
@@ -103,6 +104,10 @@ export function redraw_user(user_id) {
 
+    if (muting.is_user_muted(user_id)) {
+        return;
+    }
```
Would it be cleaner to put this check in `buddy_list.insert_or_move`?  That seems better given that the main enforcement of this behavior is in the `buddy_list` redraw logic.

It's a little awkward because for some reason `buddy_data` doesn't call its key `user_id`, but I'm not sure why we do that.

@showell thoughts?

timabbott commented on static/js/pm_conversations.js:
```diff
@@ -11,6 +12,20 @@ export function is_partner(user_id) {
+    }
+
+    return true;
+}
```
This feels pretty hacky that we're parsing the `,`s etc., so probably another TODO list item to cleanup after we merge the core behavior.

timabbott replied:
OK that's basically an existing problem with this module; I think what we actually want to do is make `conversation` objects have a `is_group` property, and maybe also a `user_ids` list of integers, that we can check, to have the parsing mostly at the boundaries of this module.

abhijeetbodas2001 replied:
I think comma separated user IDs is a convention followed throughout the codebase to denote recipients? Updating this module to use arrays make sense because I think it isn't too tied up with other modules. Shall I add a prep commit here?

timabbott replied:
I think the right way to think about is that `user_ids_string` is a really nice encoding for something we're going to put into a dictionary; and maybe it ends up being the case that for many modules, that's the only place we need to put the set of user IDs.  But in a context where we are going to iterate through the user IDs often, we may want to figure out a way to contain the parsing logic, at least to a function we call in 10 places, or perhaps some other approach.

I don't think doing something here is urgent.

abhijeetbodas2001 commented:
I will open separate, smaller PRs wherever possible for parts of this or followups.

timabbott commented:
That strategy makes sense to me!

timabbott commented on static/js/zulip.js:
```diff
@@ -0,0 +1,20 @@
+});
+
+window.home_msg_list = home_msg_list;
+window.current_msg_list = home_msg_list;
```
This looks like a failed merge conflict resolution; the code moved to `static/js/message_lists.js`.

abhijeetbodas2001 replied:
Yeah, my bad. Fixed.

timabbott commented on static/js/narrow_state.js:
```diff
@@ -340,6 +347,22 @@ export function excludes_muted_topics() {
+        narrowed_to_pms() ||
+        !active()
+    );
+}
```
It might be better to move this function to the class in `filter.js`, which has a really nice test suite where it'd be convenient to verify a new function like this.  And I think that's going to be what you want to call anyway.

timabbott replied:
(I think with that structure, we can avoid passing a new parameter when creating MessageList objects, and just check the attribute of the associated `filter` object?  Not sure.)

abhijeetbodas2001 replied:
Yes, I think so. We could initialize the `MessageListData` objects with just the related `Filter` object in that case.
There's a bunch or related functions in `narrow_state.js`, `narrowed_to_pms()` for example. Should all of those be moved to the `Filter` class?

abhijeetbodas2001 replied:
(I have added tests for this function in the current file though)

timabbott replied:
I don't think I understand this function; do we want something more like `return !narrowed_to_personal_pms` (pseudocode)?

We may also want a different name -- e.g. `shows_muted_user_messages` with the opposite meaning? Not sure, but generally when I see a function that lists 5 different cases like this, and don't have a clear explanation for why those are the 5 cases, I worry this is wrong.

(One detail you may not know is that you can generate a wide range of other possible narrows by just typing different combinations of operations into the search box; and `narrowed_to_search` specifically checks for a search keyword only)

timabbott commented on static/js/message_list_data.js:
```diff
@@ -182,12 +183,34 @@ export class MessageListData {
+            const recipient_id = Number.parseInt(recipients[0], 10);
+            return !muting.is_user_muted(recipient_id) && !muting.is_user_muted(message.sender_id);
+        });
+    }
```
I think it'd be cleaner to just do this in the same function that does topic mutes, rather than having two predicates, since it appears we're just going to compose these functions when they're called anyway.

abhijeetbodas2001 replied:
> since it appears we're just going to compose these functions when they're called anyway.

Actually, we can we can further use the fact that these are independent in the wrapper methods in `message_list.js`.
So, both of these methods will be composed only when adding messages etc to the MLD object, and only one of these will be run when we want to update the message list after receiving events.

I pushed this version here. It doesn't save on complexity/performance, but I think reads a bit better. Does that seem reasonable?

timabbott replied:
I think that makes sense to me.

timabbott replied:
See https://github.com/zulip/zulip/pull/16915/commits/28e14ffe4c799ff0e537c87cb966f7acaee17b0e#r627861439 but I think in retrospect this logic doesn't actually work.

abhijeetbodas2001 commented:
One possible simplification I have in mind is, can we remove the `excludes_muted_topic` and `excludes_muted_users` fields from `MessageList` class? There's no need to store them in that class -- we can access these anyways from the related `MessageListData` instance.

abhijeetbodas2001 commented:
I've rebased this PR, but it isn't ready for another review yet, since I haven't addressed all review comments -- I'll try to do that as soon as possible.

amanagr commented:
Is it safe to work on top of this PR?

timabbott commented:
Yes.  I think you can probably just pull out the commit(s) you want to iterate on to a separate PR; just coordinate with @abhijeetbodas2001.

abhijeetbodas2001 commented:
@amanagr Would be great if you can help with the last commit here. You can just cherry-pick it. It's independent from the rest of the PR.

I think I'll be able to update the first two commits here by tomorrow. I hope it isn't too late 😅

timabbott commented:
Sounds like a plan!  Tomorrow should be fine timing for updating the remaining commits.

timabbott commented on frontend_tests/node_tests/pm_conversations.js:
```diff
@@ -29,20 +30,29 @@ run_test("partners", () => {
         {user_ids_string: "1", max_message_id: 111},
         {user_ids_string: "15", max_message_id: 7},
     ]);
+    assert.deepEqual(pmc.recent.get_strings(), ["2,8,11", "1", "15"]);
```
Can you do a `get_strings` check just before muting these users, just so the test makes the change really clear?

abhijeetbodas2001 replied:
Done, I separated this out into its own test, along with some prep cleanup.

timabbott commented on static/js/pm_conversations.js:
```diff
@@ -11,6 +12,20 @@ export function is_partner(user_id) {
+        muting.is_user_muted(Number.parseInt(conversation.user_ids_string, 10))
+    ) {
+        return false;
+    }
```
I think for cases where it's a group PM thread, we might want effectively:
`if all(muting.is_user_muted(user_id) || is_current_user(user_id) for user_id in conversation.user_ids)`

(i.e. we also hide group PM threads if everyone involved is a muted user).

(Though we need to be careful to not have this hide PMs with yourself; worth including that case in our test cases since it's the kinda bug we could avoid now and then add with a future refactor).

abhijeetbodas2001 replied:
I've made it so that huddles with all users muted will be hidden, and also expanded the tests to include the self user.

timabbott replied:
This is great and much appreciated, thanks!

abhijeetbodas2001 commented:
@timabbott Thanks for reviewing! I've addressed most review comments, and replied inline to those which weren't clear to me. Here are the changes in the last push-
- This PR will now hide (from the PM list) not just muted users, but also huddles where all users are muted.
- Removed the `excludes_muted_topics` field from the `MessageList` class, since it was a duplicate. https://github.com/zulip/zulip/pull/16915#issuecomment-827617011
- Deleted a few redundant node tests.
- Split the "hide 1:1 PMs" commit into many smaller chunks for better readability.

---

https://github.com/zulip/zulip/pull/16915#discussion_r620748085 Suggests that we may not need these `excludes_muted_*` field in messages lists, but just the `Filter` object, right?

Re https://github.com/zulip/zulip/pull/16915#discussion_r622364669, I see that comma separated user ID strings are widely used throughout the codebase for huddles, recipients etc. Is the plan to shift to arrays overtime, or did you mean we could use arrays in just this one file (`pm_conversations.js`)?

timabbott commented on frontend_tests/node_tests/pm_conversations.js:
```diff
@@ -68,3 +70,31 @@ test("insert_recent_private_message", () => {
+        {user_ids_string: "1,2,3", max_message_id: 97},
+        {user_ids_string: "15", max_message_id: 96},
+    ]);
+    assert.deepEqual(pmc.recent.get_strings(), ["3", "1,2,3", "15"]);
```
This can be a follow-up, but it'd be nice to have a test here to ensure that "PMs with yourself" remain listed.

abhijeetbodas2001 replied:
User ID 15 is indeed the "self" user, so the last message in that array is a PM with yourself, which is listed. Am I missing something?

timabbott replied:
No, we're good, I just hadn't read the test carefully.

timabbott commented:
I merged the changes through 29bf97313a529260f1f4b95647b08fc1e48c8019, which complete the left sidebar part of this feature.  Yay!  

I think the `excludes_muted_users` check seems a bit off; 

> #16915 (comment) Suggests that we may not need these excludes_muted_* field in messages lists, but just the Filter object, right?

That's correct; I mostly suggested that because we `filter.js` node tests already setup all the cases and then let you conveniently assert which boolean properties are true for each one, with them all together, which is a nicer testing system.

abhijeetbodas2001 commented:
Earlier, I had added the `excludes_muted_users` field despite the fact that technically **all** narrows should exclude muted user PMs, considering it an optimization which will allow us to skip filtering messages in narrows *we know cannot contain PMs* (example, stream narrows which definitely cannot contain PMs will have `excludes_muted_users` as `True`, and hence we can skip some filtering work here).

This reasoning was a bit different than that for `excludes_muted_topics`, which indicates whether or not a narrow should show muted topic messages, *assuming it has some*. (So, stream narrows will have this as `True`, and search narrows as `False`).

So, to sum it up, while `excludes_muted_topics` is business logic, `excludes_muted_users` has been an optimization. These variables are in a similar vein, but I agree that this difference can be confusing.

---

I now think we have two ways forward-

#### 1. Replace the (currently incomplete) `excludes_muted_user`, with its complement.

> One detail you may not know is that you can generate a wide range of other possible narrows by just typing different combinations of operations into the search box

I wasn't aware of this, but it's definitely a problem. Because of this fact, it is not possible to have an exhaustive list of narrows which *can contain PMs*, so using the complement you describe makes sense!

> do we want something more like return !narrowed_to_personal_pms (pseudocode)?

So, we can have a new field like `can_contain_non_huddle_pms` (or something like that), which will be-

```javascript
!narrowed_to_stream() && !narrowed_to_topic()
```
because stream and topic narrows are the only place which we know definitely cannot contain PMs.

#### 2. Get rid of this "optimization" all together, and **always** filter for muted user messages.

This is in case this optimization isn't worth the trouble. Some reasons "for" this approach I could think of are-
- From what I understand, we don't store more than a few hundred messages in a message list, so I think filtering should not be too expensive. But I'm not sure.
- As described above, we will anyways need to do filtering (for muted user PMs) in all narrows except stream and stream+topic narrows, and even in stream narrows, we anyways need to filter for muted *topic* messages, so its already `O(n)`. That just leaves stream+topic narrows where this optimization will save us some work, but I'm not sure if that will be significant.
- This will make the code much simpler.

---

@timabbott Which of these options (or any other you want to suggest) make more sense?

abhijeetbodas2001 commented:
Another question I have is, should we or should we not hide muted user PMs when looking at a `pm-with/muted-user` narrow?
To specifically go to this narrow, even when you need some effort (because the muted user isn't going to be visible easily anywhere in the UI) seems like a case where you are "pulling" information, like @alya describes in https://github.com/zulip/zulip/pull/18222#issuecomment-823480927.
Moreover, showing messages there will also allow us to avoid the bug I describe in the commit message-

```
There is an edge case here. In `pm-with/muted-user`,
this commit just shows a blank narrow, without any
empty narrow message. Ideally, we should show something
like "This user is muted" and give an option to unmute.
But this cannot be done (as of now), because the narrow
isn't really "empty". The current in-code meaning of a
narrow being empty is-
(i) We should not have fetched any messages from the
server, AND
(ii) The current message list should have zero showable
messages.

In `pm-with` narrows, none of the messages (if any) are
showable even if we fetch them from the server, hence
the narrow isn't "empty" and we don't show any empty
narrow message. The solution here could be either
(1) filtering muted messages in the backend itself or
(2) keep fetching messages till we receive some unmuted
messages (or end up fetching all of the narrow messages,
in case all are muted) and disregard (i) from above.

This edge case can likely be ignored for now, until
we decide which of the above to proceed with.
Also, it would take some effort for muters to reach
a `pm-with/mutee` narrow, because we will be hiding
mutees from buddy list and also be hiding messages
sent by mutees.

A similar bug is present in stream narrows due to
topic muting. This happens when all the recent messages
in a stream are not showable, and we are on a stream
narrow. See #8733.
```

timabbott commented:
I think it'd be reasonable for pm-with narrows to a single muted user to have every message the muted user sent by hidden, and the message you sent appear as normal -- basically the same behavior we'd have in a group PM thread if you explicitly navigated to it.

abhijeetbodas2001 commented:
https://user-images.githubusercontent.com/55339528/117146037-689fb800-add1-11eb-95c2-92ca27aa32de.mp4

abhijeetbodas2001 commented on static/js/message_list_data.js:
```diff
@@ -178,6 +178,18 @@ export class MessageListData {
+            // Actually these will never be undefined, but this condition makes it possible for
+            // our test suite modules (which are testing things other than muting) to create message
+            // lists using dummy or undefined filter objects.
+            return [...messages];
```
This part is a bit hacky. The need to do this is other tests involving message lists fail with this change if they haven't defined a `filter` object for the list.
I thought of stubing `message_list_data.messages_filtered_for_user_mutes` instead -- would that be better?

timabbott replied:
How many tests use this trick?  If it's not hard, the best option would probably be for message_list objects in tests to always have a filter object, if it's only a bit of stubbing work?  Second best is the stubbing approach.

abhijeetbodas2001 replied:
Turns out not many. It was actually a nice opportunity to remove some unnecessary mocks 🙂. I added a prep commit to do so.

timabbott commented:
I merged 3/5 commits; the only thing remaining in this PR is the "hide from the UI 1:1 PMs entirely".  Can you talk about that in the #design thread?  I'd like to make sure we've discussed whether we want that behavior.

timabbott commented on static/js/muting_ui.js:
```diff
@@ -157,9 +157,9 @@ export function unmute_user(user_id) {
     if (message_lists.current !== message_lists.home) {
-        message_lists.home.rerender();
+        message_lists.home.update_user_muting_and_rerender();
     }
```
This can be a follow-up, but we should probably do a refactor to make this pattern, which have in like 20 places, not involve duplication.  E.g. 
```
message_lists.for_all_lists((message_list) => message_list.update_muting_and_rerender());
```

@ryanreh99 FYI in case you have thoughts on the precise interface.

timabbott commented on static/js/message_list_data.js:
```diff
@@ -182,15 +177,37 @@ export class MessageListData {
+    }
+
+    update_items_for_user_muting() {
+        this._items = this.messages_filtered_for_user_mutes(this._all_items);
```
It seems wrong that both of these functions start from `_all_items` and store that to `_items` -- won't that make calling one or the other oscillate between whether topic or user muting is respected?

timabbott replied:
I think the conclusion is we just need an `update_items_for_muting` function that gets run for both updates.

abhijeetbodas2001 replied:
> It seems wrong that both of these functions start from _all_items and store that to _items -- won't that make calling one or the other oscillate between whether topic or user muting is respected?

This is a blunder. Really sorry about that. The tests did not catch this because the only tested the behavior of the `MessageListData.unmuted_messages` (which returns filtered messages) and not what is stored in `_items`

abhijeetbodas2001 replied:
I've made it so that we now have a single function to update messages after muting events, and also added another assertion to check that `_items` stores the correct set of messages.

timabbott replied:
Sounds good.  Yeah, tests can only do so much: you have to reason through how something will work.  Which is one reason code review is really important: Having two people make sure the logic makes sense to them avoids a lot of potential bugs.

timabbott commented on static/js/message_list_data.js:
```diff
@@ -178,6 +178,18 @@ export class MessageListData {
     }
 
     messages_filtered_for_user_mutes(messages) {
+        if (this.filter === undefined || this.filter.is_non_huddle_pm === undefined) {
```
is `is_non_huddle_pm` a function or a property?

abhijeetbodas2001 replied:
It is a function. I've removed this check now, and fixed a few tests to use real `Filter` objects instead.

timabbott commented:
OK, posted what I think is a complete set of comments on this.  I'm really excited that this is almost done!

The last thing that remains, which we should do in a new PR because this one is too long, is a /help/mute-users article on user muting.  @alya can help with editing for this, but we should try to transcribe the semantics from our design document into a somewhat user-friendly page that explains muting, unmuting, etc.

And we should include in that PR the commit to publish the feature (remove the development-only guards).

https://help.twitter.com/en/using-twitter/twitter-mute and the similar article for blocking are probably good samples for how to talk about these things in a non-technical way.

abhijeetbodas2001 commented:
@timabbott Thanks for the review! I've addressed your comments. This is ready for another look.
I'll also open the PR for the /help page soon.

timabbott commented:
Awesome, merged, thanks @abhijeetbodas2001!
